jax/interpreters/xla.py

```diff
@@ -149,23 +149,40 @@ def _make_abstract_python_scalar(typ, _):
 
 def apply_primitive(prim, *args, **params):
   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
-  abstract_args = map(abstractify, args)
-  compiled_fun = xla_primitive_callable(prim, *abstract_args, **params)
+  compiled_fun = xla_primitive_callable(prim, *map(arg_spec, args), **params)
   return compiled_fun(*args)
 
+def arg_spec(x):
+  aval = abstractify(x)
+  try:
+    return aval, x._device
+  except:
+    return aval, None
+
 @cache()
-def xla_primitive_callable(prim, *abstract_args, **params):
+def xla_primitive_callable(prim, *arg_specs, **params):
   backend = params.get('backend', None)
-  aval_out = prim.abstract_eval(*abstract_args, **params)
+  avals, devices = unzip2(arg_specs)
+  # TODO(mattjj): make Device hashable instead of handling pairs here
+  try:
+    device, = set(d for d in devices if d is not None) or (None,)
+  except ValueError:
+    msg = "primitive arguments must be colocated on the same device, got {}"
+    names = ("{}({})".format(d[0].__name__, d[1]) for d in devices if d is not None)
+    raise ValueError(msg.format(", ".join(names)))
+  else:
+    all_devices = it.chain(xb.devices(), xb.devices('cpu'))
+    device = device and next(d for d in all_devices if (type(d), d.id) == device)
+  aval_out = prim.abstract_eval(*avals, **params)
   if prim.multiple_results:
     handlers = tuple(map(aval_to_result_handler, aval_out))
     handle_result = lambda xs: tuple(h(x) for h, x in zip(handlers, xs.destructure()))
   else:
     handle_result = aval_to_result_handler(aval_out)
-  tuple_args = len(abstract_args) > 100
-  built_c = primitive_computation(prim, tuple_args, *abstract_args, **params)
-  compiled = built_c.Compile(compile_options=xb.get_compile_options(),
-                             backend=xb.get_backend(backend))
+  tuple_args = len(avals) > 100
+  built_c = primitive_computation(prim, tuple_args, *avals, **params)
+  options = xb.get_compile_options(device_assignment=(device.id,) if device else None)
```

Comment:
Since XLA doesn't specialize the compiled program for a particular device ID, we maybe don't even need to know the common device in Python at all--we could simply compile it for a single device, then let XLA check that the buffers are all located on the same device and launch with the right device_assignment.

Reply:
Are you proposing something XLA could change in the future, or proposing something we could do differently now? Because as it stands the XLA program needs to be compiled for the device to which the argument buffers belong. That is, if we don't specify the device id as a compile option here, we get errors.

Reply:
The XLA Python bindings require the device id to be specified, but XLA itself doesn't (unless I'm misreading https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/python/local_client.cc#L982-L988, the only thing the Python bindings do with it at compile time is to set it as a property of the resulting computation; it's only used at runtime).

Reply:
Sounds like a good future change to make in the XLA client!

Reply:
But in a future PR, possibly by someone who understands that code better ;)